controllers/etcdbackup_controller.go

```diff
@@ -16,16 +25,169 @@ type EtcdBackupReconciler struct {
 	Log logr.Logger
 }
 
-// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdbackups,verbs=get;list;watch;create;update;patch;delete
+func (r *EtcdBackupReconciler) reconcile(
+	ctx context.Context,
+	resourceID string,
+	resource *etcdv1alpha1.EtcdBackup) (*etcdv1alpha1.EtcdBackupPhase, string, error) {
+
+	// If status is not yet set, the resource was probably just created.
+	if resource.Status.Phase == "" {
+		// Mark the backup as started.
+		return &etcdv1alpha1.EtcdBackupPhaseBackingUp, "", nil
+	}
+
+	// If the backup has finished, or failed, don't take any action.
+	if resource.Status.Phase == etcdv1alpha1.EtcdBackupPhaseCompleted || resource.Status.Phase == etcdv1alpha1.EtcdBackupPhaseFailed {
+		return nil, "", nil
+	}
+
+	// The backup will live on disk in a well-known location, unique to the resource describing this backup.
+	localPath := filepath.Join(os.TempDir(), string(resource.UID), "snapshot.db")
+
+	// Extract backup method config.
+	method := &backup.SnapshotMethod{
+		Endpoints: func() []string {
+			s := make([]string, len(resource.Spec.ClusterEndpoints))
+			for i, e := range resource.Spec.ClusterEndpoints {
+				s[i] = (&url.URL{
+					Scheme: strings.ToLower(string(e.Scheme)),
+					Host:   fmt.Sprintf("%s:%d", e.Host, e.Port.IntVal),
+				}).String()
+			}
+			return s
+		}(),
```

Comment:
Why do we inline and then execute a function here?

Reply:
just saves pouting the scope with single-use variables/functions. doing that is is a really weakly held opinion so i've changed it to something more conventional